File: src/python/generator/lib/hlcsnowflake/hlc_snowflake.py
import time
import threading
from lib.network_util import get_node_id_from_ip

class HlcSnowflake:
    def __init__(self):
        self.EPOCH = 1767225600000 # Jan 1, 2026
        self.NODE_ID_BITS = 10
        self.SEQUENCE_BITS = 12
        self.MAX_NODE_ID = (1 << self.NODE_ID_BITS) - 1
        self.MAX_SEQUENCE = (1 << self.SEQUENCE_BITS) - 1
        self.NODE_ID_SHIFT = self.SEQUENCE_BITS
        self.TIMESTAMP_SHIFT = self.SEQUENCE_BITS + self.NODE_ID_BITS

        self.node_id = get_node_id_from_ip() & self.MAX_NODE_ID
        self.logical_time = 0
        self.sequence = 0
        self.lock = threading.Lock()

    def physical_time_millis(self):
        return int(time.time() * 1000)

    def next_id_string(self):
        with self.lock:
            physical_time = self.physical_time_millis()
            current_logical_time = self.logical_time

            if physical_time > current_logical_time:
                new_logical_time = physical_time
                self.logical_time = new_logical_time
                self.sequence = 0
            else:
                new_logical_time = current_logical_time
                self.sequence = (self.sequence + 1) & self.MAX_SEQUENCE
                if self.sequence == 0:
                    new_logical_time = self.logical_time + 1
                    self.logical_time = new_logical_time

            id = ((new_logical_time - self.EPOCH) << self.TIMESTAMP_SHIFT) | \
                 (self.node_id << self.NODE_ID_SHIFT) | \
                 self.sequence

            return str(id)

File: src/go/generator/lib/hlcsnowflake/hlc_snowflake.go
package hlcsnowflake

import (
	"strconv"
	"sync/atomic"
	"time"

	"github.com/dilipkumardk/uuid-generator/src/go/generator/lib"
)

const (
	epoch          = uint64(1767225600000) // Jan 1, 2026
	nodeIdBits     = 10
	sequenceBits   = 12
	maxNodeId      = (1 << nodeIdBits) - 1
	maxSequence    = (1 << sequenceBits) - 1
	nodeIdShift    = sequenceBits
	timestampShift = sequenceBits + nodeIdBits
)

type HlcSnowflake struct {
	nodeId      uint64
	logicalTime atomic.Uint64
	sequence    atomic.Uint64
}

func New() *HlcSnowflake {
	return &HlcSnowflake{
		nodeId: lib.GetNodeIdFromIp() & maxNodeId,
	}
}

func (s *HlcSnowflake) physicalTimeMillis() uint64 {
	return uint64(time.Now().UnixMilli())
}

func (s *HlcSnowflake) NextId() uint64 {
	physicalTime := s.physicalTimeMillis()
	currentLogicalTime := s.logicalTime.Load()

	var newLogicalTime uint64

	if physicalTime > currentLogicalTime {
		newLogicalTime = physicalTime
		s.logicalTime.Store(newLogicalTime)
		s.sequence.Store(0)
	} else {
		newLogicalTime = currentLogicalTime
		seq := (s.sequence.Add(1)) & maxSequence
		if seq == 0 {
			newLogicalTime = s.logicalTime.Add(1)
		}
	}

	id := ((newLogicalTime - epoch) << timestampShift) |
		(s.nodeId << nodeIdShift) |
		s.sequence.Load()

	return id
}

func (s *HlcSnowflake) NextIdString() string {
	return strconv.FormatUint(s.NextId(), 10)
}
